### PR TITLE
Programmatically add emails to Fawkes users

### PR DESCRIPTION
**Pull request by JordanGuinn**
## Background
This PR updates all scripts and code bits responsible for creating/updating user data to now account for emails coming directly from WingedKeys, so we can keep them in Fawkes as well (given WingedKeys will eventually be replaced with ForgeRock).

## GitHub Issue
https://github.com/ctoec/data-collection/issues/1313

## Associated PRs
https://github.com/ctoec/winged-keys/pull/74

## Validation Plan
- [x] Confirm users programmatically created on first login in non-prod environments now have emails added

## Automated Testing
- [x] All unit tests are passing.
- [x] All e2e tests are passing.

## Additional Tasks
- [x] Add emails to all users in `devsecure`
- [x] Add emails to all users in `prod`

### CODE REVIEW

**Review by bamader — APPROVED** [review]
God these tokens are ugly 😭

**Review by vonbearshark — APPROVED** [review]

File: src/data/prod/users/upsert.ts [hunk 1]
```diff
@@ -81,9 +83,27 @@ export const createUserData = async (
...
         createdCount += 1;
       } else {
-        console.log(
-          `\tUser ${row.userName} with username ${row.email} already exists`
-        );
+        let updatedUser: Partial<User> = {};
+
+        if (!!_firstName && _firstName !== user.firstName)  updatedUser.firstName = _firstName;
+        if (!!_lastName && _lastName !== user.lastName)  updatedUser.lastName = _lastName;
+        if (!!row.email && row.email !== user.email)  updatedUser.email = row.email;
```
Comment (vonbearshark) [inline src/data/prod/users/upsert.ts:90]: nit: can we do: `row?.email !== user.email`, etc?

File: src/middleware/authenticate.ts [hunk 2]
```diff
@@ -112,7 +139,7 @@ const getUser = async (wingedKeysId: string) => {
  */
 async function getUserFromWingedKeys(
   bearerToken: string
-): Promise<AxiosResponse<any>> {
+): Promise<AxiosResponse<UserInfoResponse>> {
```
Comment (vonbearshark) [inline src/middleware/authenticate.ts:142]: 👏🏻

File: src/middleware/authenticate.ts [hunk 3]
```diff
@@ -37,7 +62,9 @@ const decodeClaim = jwt({
  */
 const addUser = passAsyncError(
   async (req: Request, _: Response, next: NextFunction) => {
-    if (req.claims?.sub) {
+    let claims = (req.claims || {}) as TokenClaims;
```
Comment (vonbearshark) [inline src/middleware/authenticate.ts:65]: nit nit: `const` and `??` (instead of `||`)

**Review by vonbearshark — APPROVED** [review]

File: src/data/prod/users/upsert.ts (hunk 1 above)
Comment (JordanGuinn) [reply src/data/prod/users/upsert.ts:90]: That would just confirm that `row` is defined, not necessarily that `email` is defined, yeah?  Can you use that conditional `?` check on primitives?

**Comment by jlhogan** [comment]
so the approach here is that we're updating user data in ECE Reporter from a file? 
how are we imagining we manage updating emails in wingedkeys? it appears we can get emails in the tool from the user claim, so is there realy a need to store emails in ECE Reporter? As long as the emails are up to date in wingedkeys DB (with correct id FKs to ECE Reporter user table) replacing ID server with forgerock should still be easily achievable. 

if we can get email from ID Server in the claim, then i think storing emails in the tool just introduces an additional opportunity for things to get out of sync. But can we actually get that custom claim from ID Server?

**Comment by JordanGuinn** [comment]
We can get the _currently authenticated_ user's email from the claims we decode from the JWT, but that's wildly different from getting all emails of all users in WingedKeys.  We _could_ do that, it would just be an unknown amount of unprecedented, throwaway work (because of the ForgeRock shift).

Re the email update question: I haven't put any thought towards non-manually updating emails in WingedKeys.  Is that a thing we're gonna need to provide programmatic capabilities for?  How often would that be necessary?

**Comment by jlhogan** [comment]
duhhhhhhhhhhhhhhhhhhhh wow my brain. that's what i get for early AM reviews.

Re "updating" emails in wingedkeys .. i think what i mean more than that is actually this:
would it be useful to have a process that reconciles emails + ids in wingedkeys with emails + wingedkeysid in ece reporter user table? that way we just make the updates in wingedkeys when we hear them, and then run a script that updates any mismatched emails in the ECE Reporter DB. seems slightly less error prone than:
1. get a list with updated user emails
2. update emails in wingedkeys
3. run process to update emails in ece reporter

because if a user email changes it DEFINITELY needs to be changed in wingedkeys since that's how they sign on/ get new password

**Comment by JordanGuinn** [comment]
Ohhhhhh, I think I follow.  In my head, I had #1 and #2 reversed, where the list of emails used to update Fawkes was only ever exported _directly_ from the WingedKeys database.  I assumed that we weren't going to allow admins to programmatically update users' emails through Fawkes _at all_, so this script would only ever need to be one-way (WingedKeys => Fawkes).

So the process for updating a user's email, should we ever need to do that, would be:

1. Manually update said email in WingedKeys
2. Export appropriate WingedKeys users table columns to file (probably worth documenting)
3. Run script to update Fawkes from said file

Which I think is effectively what you're describing as well?

**Comment by JordanGuinn** [comment]
I now see the script does both WingedKeys user creates AND Fawkes user creates at the same time 😅 Will split the two, so we're able to do one without the other

File: src/data/prod/users/update.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  });
+
+  for (const user of usersToUpdate) {
+    try {
+      const wingedKeysUser: BasicWingedKeysUser = wkUsersById[user.id];
+      if (!wingedKeysUser) {
+        throw new Error('This is actually impossible');
+      }
+
+      const [firstName, lastName] = wingedKeysUser.userName.split(' ');
```
Comment (jlhogan) [inline src/data/prod/users/update.ts:37]: we should not do this. only very fiew wingedkeys users have usernames that are actual username. In most cases it's the email. also, have updated some firstname/lastnames in ECE db and not in wingedkeys db. TLDR; wingedkeys is source of truth for email addresses (which should = usernames, except in a few legacy cases) and ECE reporter DB is source of truth for names.

File: src/data/prod/users/update.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+      let updatedUser: Partial<User> = {};
+
+      if (!!firstName && firstName !== user.firstName)
+        updatedUser.firstName = firstName;
+      if (!!lastName && lastName !== user.lastName)
+        updatedUser.lastName = lastName;
+      if (!!wingedKeysUser.email && wingedKeysUser.email !== user.email)
```
Comment (jlhogan) [inline src/data/prod/users/update.ts:45]: as mentioned above, imo we should only be checking/updated emails not firstname/lastname

File: src/data/prod/users/update.ts (hunk 4 above)
Comment (JordanGuinn) [reply src/data/prod/users/update.ts:37]: Works for me - this was just a direct pull from the create script anyway

**Pull request merged**